ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -1275,7 +1275,8 @@ FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws St
         // recheck to make sure that offsetToStart hasn't slipped out of the journal. It is possible (but highly
         // improbable) that the offset has slipped out of the journal and has been compacted b/w the time of getting
         // entries from the journal to when another view of the index segments was obtained.
-        if (entries != null && offsetToStart.compareTo(journal.getFirstOffset()) >= 0) {
+        if (entries != null && journal.getFirstOffset() != null
+            && offsetToStart.compareTo(journal.getFirstOffset()) >= 0) {
```

Comment:
this would not fix the race condition, we should probably do 
```
Offset journalFirstOffset = journal.getFirstOffset();
if (entries != null && journalFirstOffset != null && offsetToStart.compareTo(journalFirstOffset)) >= 0) {
}
```

Reply:
Sure, Updated it and updated other parts which call journal.getFirstOffset() twice.